molgenis-app-manager/src/main/java/org/molgenis/app/manager/service/impl/AppManagerServiceImpl.java

```diff
@@ -0,0 +1,261 @@
+		String pluginId = generatePluginId(app);
+		Plugin plugin = pluginFactory.create(pluginId);
+		plugin.setLabel(app.getLabel());
+		plugin.setDescription(app.getDescription());
+		dataService.add(PluginMetadata.PLUGIN, plugin);
+	}
+
+	@Override
+	@Transactional
+	public void deactivateApp(String id)
+	{
+		App app = findAppById(id);
+		app.setActive(false);
+		dataService.update(AppMetadata.APP, app);
+
+		String pluginId = generatePluginId(app);
+		dataService.deleteById(PluginMetadata.PLUGIN, pluginId);
+
+		// TODO remove from menu JSON?
+	}
+
+	@Override
+	@Transactional
+	public void deleteApp(String id) throws IOException
+	{
+		App app = findAppById(id);
+		deleteDirectory(new File(app.getResourceFolder()));
+		dataService.deleteById(AppMetadata.APP, id);
+	}
+
+	@Override
+	@Transactional
+	public void uploadApp(MultipartFile multipartFile) throws IOException, ZipException
+	{
+		String appArchiveName = "zip_file_" + multipartFile.getOriginalFilename();
+		ZipFile appArchive = new ZipFile(fileStore.store(multipartFile.getInputStream(), appArchiveName));
+
+		if (!appArchive.isValidZipFile())
+		{
+			fileStore.delete(appArchiveName);
+			throw new InvalidAppArchiveException(multipartFile.getName());
+		}
+
+		String appDirectoryName = fileStore.getStorageDir() + File.separator + multipartFile.getOriginalFilename();
+		appArchive.extractAll(appDirectoryName);
+		fileStore.delete(appArchiveName);
+
+		checkForMissingFilesInAppArchive(appDirectoryName);
+
+		File indexFile = new File(appDirectoryName + File.separator + ZIP_INDEX_FILE);
+		File configFile = new File(appDirectoryName + File.separator + ZIP_CONFIG_FILE);
+		if (!isConfigContentValidJson(configFile))
+		{
+			fileStore.deleteDirectory(appDirectoryName);
+			throw new InvalidAppConfigException();
+		}
+
+		AppConfig appConfig = gson.fromJson(fileToString(configFile), AppConfig.class);
+		checkForMissingParametersInAppConfig(appConfig, appDirectoryName);
+
+		App newApp = appFactory.create();
+		newApp.setLabel(appConfig.getLabel());
+		newApp.setDescription(appConfig.getDescription());
+		newApp.setAppVersion(appConfig.getVersion());
+		newApp.setApiDependency(appConfig.getApiDependency());
+		newApp.setTemplateContent(fileToString(indexFile));
+		newApp.setActive(false);
+		newApp.setIncludeMenuAndFooter(appConfig.getIncludeMenuAndFooter());
+		newApp.setResourceFolder(appDirectoryName);
+
+		// If provided config does not include runtimeOptions, set an empty map
+		Map<String, Object> runtimeOptions = appConfig.getRuntimeOptions();
+		if (runtimeOptions == null) runtimeOptions = Maps.newHashMap();
+		newApp.setAppConfig(gson.toJson(runtimeOptions));
+
+		newApp.setUri(appConfig.getUri());
+		dataService.add(AppMetadata.APP, newApp);
+	}
+
+	private String generatePluginId(App app)
+	{
+		String pluginId = APP_PLUGIN_ROOT + app.getUri();
+		if (!pluginId.endsWith("/"))
+		{
+			pluginId = pluginId + "/";
+		}
+		return pluginId;
+	}
+
+	private App findAppById(String id)
```

Comment:
this method returns the app ( or throws an exception) name should be 'getAppById' or 'getApp'

Reply:
PR